src/newgrf_engine.cpp

```diff
@@ -1328,44 +1330,35 @@ static bool EnginePreSort(const EngineID &a, const EngineID &b)
  */
 void CommitVehicleListOrderChanges()
 {
+	/* Build a list of EngineIDs. EngineIDs are sequential from 0 up to the number of pool items with no gaps. */
+	std::vector<EngineID> ordering(Engine::GetNumItems());
+	std::iota(std::begin(ordering), std::end(ordering), 0);
+
 	/* Pre-sort engines by scope-grfid and local index */
-	std::vector<EngineID> ordering;
-	for (const Engine *e : Engine::Iterate()) {
-		ordering.push_back(e->index);
-	}
-	std::sort(ordering.begin(), ordering.end(), EnginePreSort);
+	std::ranges::sort(ordering, EnginePreSort);
 
 	/* Apply Insertion-Sort operations */
-	for (const ListOrderChange &it : _list_order_changes) {
-		EngineID source = it.engine;
-		uint local_target = it.target;
+	for (const ListOrderChange &loc : _list_order_changes) {
+		EngineID source = loc.engine;
 
-		const EngineIDMapping *id_source = _engine_mngr.data() + source;
-		if (id_source->internal_id == local_target) continue;
+		const EngineIDMapping &id_source = _engine_mngr[source];
+		if (id_source.internal_id == loc.target) continue;
 
-		EngineID target = _engine_mngr.GetID(id_source->type, local_target, id_source->grfid);
+		EngineID target = _engine_mngr.GetID(id_source.type, loc.target, id_source.grfid);
 		if (target == INVALID_ENGINE) continue;
 
-		int source_index = find_index(ordering, source);
-		int target_index = find_index(ordering, target);
+		auto it_source = std::ranges::find(ordering, source);
+		auto it_target = std::ranges::find(ordering, target);
 
-		assert(source_index >= 0 && target_index >= 0);
-		assert(source_index != target_index);
+		assert(it_source != std::end(ordering) && it_target != std::end(ordering));
+		assert(it_source != it_target);
 
-		EngineID *list = ordering.data();
-		if (source_index < target_index) {
-			--target_index;
-			for (int i = source_index; i < target_index; ++i) list[i] = list[i + 1];
-			list[target_index] = source;
-		} else {
-			for (int i = source_index; i > target_index; --i) list[i] = list[i - 1];
-			list[target_index] = source;
-		}
+		if (it_target < it_source) std::rotate(it_target, it_source, std::next(it_source));
+		if (it_source < it_target) std::rotate(it_source, std::next(it_source), it_target);
```

Comment:
Is it me, or is this logic somewhat misleading?
Given `assert(it_source != it_target)`, either `it_target < it_source` or `it_source < it_target` is valid. As far as I am aware `std::rotate` does not update the passed iterators. So isn't this actually an if-else?
Or is this to work around some UB that would be triggered when assertions are disabled?